elastic/kibana
# [Security solution] Implement dashboard to track Gen AI Token Usage

## Summary

Resolves https://github.com/elastic/security-team/issues/6712

1. Implements new gen ai subaction called `getDashboard`. The subaction checks if a dashbord with a given id exists, and if not creates the dashboard. Returns `{ exists: true }` upon successful retreival/creation
2. The subaction also checks the users permissions on the index pattern used by the dashboard, and will return  `{ exists: false }` if the user does not have permissions.
3. When the `getDashboard` subaction returns  `{ exists: true }` and  `isEdit` is true, the connector form displays a link to the dashboard filtered by the connector id
<img width="925" alt="Screenshot 2023-06-05 at 5 17 46 PM" src="https://github.com/elastic/kibana/assets/6935300/3cd8ed8e-a109-40b2-9ca0-6e638feeb7da">
<img width="1439" alt="Screenshot 2023-06-05 at 5 21 09 PM" src="https://github.com/elastic/kibana/assets/6935300/ec88c922-9cf1-4335-9875-b68968f9128d">

## Type confusion

It seems the `stack_connectors` uses the `triggers_actions_ui`  `CONNECTORS_PLUGIN_ID` which has its own `TriggersAndActionsUiServices` type. However, when I import `useKibana` it is tied to the type for `PLUGIN_ID`. That is why I extended both versions of the app to include `dashboard`. Please let me know if you think I should instead maybe export a `useKibana` that is specifically typed to the connector?

### To test
Credentials for OpenAI and Azure OpenAI: https://p.elstc.co/paste/+6nHnsHt#WJOOGIkVoJjTq7zPLP77BzSG1AUxNFUDvcnSfLPtO7m

1. Create at least two GenAI connectors with OpenAI or Azure OpenAI credentials
2. Create multiple users
3. Log in with each user and send a variety of prompts using the "Test" panel for each connector
5. Open one of the connectors and click the new link to dashboard
6. Confirm that the different users token usage is reflected
7. Confirm that the view is filtered to the connector id:
<img width="994" alt="Screenshot 2023-06-05 at 5 29 50 PM" src="https://github.com/elastic/kibana/assets/6935300/5e4b236a-770a-44be-bace-bf265933617a">
8. Remove the filter, confirm that the Lens charts reflect that change and shows usage for both of your connectors

### Testing permissions
1. Create a user with the "Viewer" role. 
2. Login with that user and open a Gen AI connector edit view in stack management
3. Notice that the link to the dashboard is not there


_**Update 6/12/23:** Updated to selectively show the link to the dashboard based on user permissions_
_**Update 6/13/23:** Added a markdown text block to explain dashboard permissions, content to be updated with a doc link_
<img width="1410" alt="Screenshot 2023-06-13 at 11 09 58 AM" src="https://github.com/elastic/kibana/assets/6935300/ae3ebf60-a354-4743-893f-03dd76ed6a0e">
